### PR TITLE
Fix CodeMetrics source "LineCount" metric to not include leading and …

### PR DESCRIPTION
**Pull request by mavasani**
…trailing empty newlines.

Fixes customer feedback from https://developercommunity.visualstudio.com/content/problem/1014036/code-metrics-lines-of-source-code-off-by-one.html

### CODE REVIEW

File: src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs [hunk 1]
```diff
@@ -86,11 +86,42 @@ internal static async Task<long> GetLinesOfCodeAsync(ImmutableArray<SyntaxRefere
...
+                    for (; loopCondition(index, triviaParts.Length); index += incrementOrDecrement, count++)
+                    {
+                        var trivia = triviaParts[index].Trim();
+                        if (trivia.Length != 0)
+                            break;
+                    }
+
+                    return count;
+
+                    static bool LoopConditionForLeading(int index, int length) => index < length - 1;
```
Comment (genlu) [inline src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs:121]: This is a local function to a local function to a local function :D

File: src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs [hunk 2]
```diff
@@ -86,11 +86,42 @@ internal static async Task<long> GetLinesOfCodeAsync(ImmutableArray<SyntaxRefere
...
+                static int GetNewlineCount(ImmutableArray<string> triviaParts, bool leading)
+                {
+                    var index = leading ? 0 : triviaParts.Length - 1;
+                    var loopCondition = leading ? LoopConditionForLeading : (Func<int, int, bool>)LoopConditionForTrailing;
+                    var incrementOrDecrement = leading ? 1 : -1;
+                    var count = 0;
+                    for (; loopCondition(index, triviaParts.Length); index += incrementOrDecrement, count++)
+                    {
+                        var trivia = triviaParts[index].Trim();
+                        if (trivia.Length != 0)
```
Comment (dotpaul) [inline src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs:115]: >Length [](start = 35, length = 6)

Do comments count as syntax trivia? Didn't see any unit tests with comments in the test code.

Comment (mavasani) [reply src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs:115]: Yes, good point, let me add a test.

File: src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs (hunk 1 above)
Comment (mavasani) [reply src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs:121]: Yep and all static - I like such static nesting :)

File: src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs (hunk 2 above)
Comment (genlu) [inline src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs:115]: `if (!string.IsNullOrWhiteSpace(triviaParts[index])) {} `?

File: src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs [hunk 3]
```diff
@@ -86,11 +86,42 @@ internal static async Task<long> GetLinesOfCodeAsync(ImmutableArray<SyntaxRefere
...
+                            break;
```
Comment (genlu) [inline src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs:116]: A couple of questions:
1. Looks like we count comments as part of "Lines"?
2. Do we want to exclude empty newlines in between comments too?

File: src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs (hunk 2 above)
Comment (mavasani) [reply src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs]: Yep, helped simplify the loop. Thanks!

File: src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs (hunk 3 above)
Comment (mavasani) [reply src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs]: I thought about it, but I think it is reasonable to count the start of any non-whitespace trivia as a line belonging to next declaration.

Comment (dotpaul) [reply src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs]: Whatev the answer is, should have a test with blank lines between comments.

Comment (mavasani) [reply src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs]: Sure, added test. We can go both ways here, but this seems to be a rare case, so we can revisit if we get feedback this is problematic.

**Pull request merged**